# Required imports

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import plotly.express as px
import json
from ipywidgets import interact
import plotly.graph_objects as go

# 

# Data preparation

In [2]:
data = pd.read_csv('us events ad_revenue filtered 03.02-07.02.csv')
data.head()

profile_id os_version device_manufacturer   device_model device_type  \
0  JLjpGzDHBQm         11             Samsung  Galaxy Tab S6      tablet   
1  JLjpGzDHBQm         11             Samsung  Galaxy Tab S6      tablet   
2  JLkVacrzHY6         11              Google        Pixel 2       phone   
3  JMC6wG3cX0f         10             Samsung      Galaxy S9       phone   
4  JLzWWTJfi1b         10             Samsung      Galaxy S9       phone   

  device_locale    event_name  \
0         en_US  c_ad_revenue   
1         en_US  c_ad_revenue   
2         en_US  c_ad_revenue   
3         en_US  c_ad_revenue   
4         en_US  c_ad_revenue   

                                          event_json       event_datetime  \
0  {"ad_type":"banner","placement":"","view_index...  2022-02-04 20:02:28   
1  {"ad_type":"banner","placement":"","view_index...  2022-02-04 20:02:45   
2  {"ad_type":"banner","placement":"","view_index...  2022-02-05 17:53:05   
3  {"ad_type":"int","placement":"next_level","vie...  2022-02-06 17:09:01   
4  {"ad_type":"banner","placement":"","view_index...  2022-02-06 17:15:38   

   event_timestamp event_receive_datetime  event_receive_timestamp  \
0       1643994148    2022-02-06 17:58:19               1644159499   
1       1643994165    2022-02-06 17:58:19               1644159499   
2       1644072785    2022-02-06 16:59:29               1644155969   
3       1644156541    2022-02-06 17:09:24               1644156564   
4       1644156938    2022-02-06 17:16:16               1644156976   

  connection_type operator_name country_iso_code       city  
0            wifi           NaN               US    Norwalk  
1            wifi           NaN               US    Norwalk  
2            wifi       Verizon               US  Charlotte  
3            wifi      T-Mobile               US      Cocoa  
4            cell      Verizon                US        NaN

In [3]:
data_json = pd.DataFrame(data['event_json'].apply((json.loads)).values.tolist())
data_json.head()

ad_type   placement  view_index  session_view_index           cur_ticks  \
0  banner                       1                   0  637795729492245130   
1  banner                       1                   0  637795729659949790   
2  banner                       1                   0  637796515863735930   
3     int  next_level           1                   1  637797353426827330   
4  banner                       1                   0  637797357392197840   

   session_length   revenue   ads_ltv   ecpm_sma  normalized_ecpm_sma  \
0       705.88070  0.000510  0.487877   0.511508             0.511508   
1       722.65120  0.000510  0.488387   0.511389             0.511389   
2       431.81290  0.002386  0.678479   0.928597             0.928597   
3       125.94960  0.013894  0.013894  13.782633            13.893784   
4        12.71129  0.006828  0.060571   1.660282             1.660282   

      value currency ad_country_code       network_name        ad_unit_id  \
0  0.000510      USD              US       Google AdMob  64c6b1114ea8c1cc   
1  0.000510      USD              US       Google AdMob  64c6b1114ea8c1cc   
2  0.002386      USD              US  APPLOVIN_EXCHANGE  64c6b1114ea8c1cc   
3  0.013894      USD              US           AppLovin  3ab7705b07ebb75f   
4  0.006828      USD              US       Google AdMob  64c6b1114ea8c1cc   

                     ad_network_placement                  creativeId  
0  ca-app-pub-7511632782617496/8825263691  CJ3XirfD5vUCFYPnKAUdG3MOYQ  
1  ca-app-pub-7511632782617496/8825263691  CNa-lL_D5vUCFZrpKAUdU2QBDQ  
2                          banner_regular                  pandora-58  
3                           inter_regular                    13045364  
4  ca-app-pub-7511632782617496/6197324251  CMTi5e-h6_UCFUdVwQodt9EIMw

In [4]:
data = pd.concat([data.drop('event_json', axis=1), data_json], axis=1)
data.head()

profile_id os_version device_manufacturer   device_model device_type  \
0  JLjpGzDHBQm         11             Samsung  Galaxy Tab S6      tablet   
1  JLjpGzDHBQm         11             Samsung  Galaxy Tab S6      tablet   
2  JLkVacrzHY6         11              Google        Pixel 2       phone   
3  JMC6wG3cX0f         10             Samsung      Galaxy S9       phone   
4  JLzWWTJfi1b         10             Samsung      Galaxy S9       phone   

  device_locale    event_name       event_datetime  event_timestamp  \
0         en_US  c_ad_revenue  2022-02-04 20:02:28       1643994148   
1         en_US  c_ad_revenue  2022-02-04 20:02:45       1643994165   
2         en_US  c_ad_revenue  2022-02-05 17:53:05       1644072785   
3         en_US  c_ad_revenue  2022-02-06 17:09:01       1644156541   
4         en_US  c_ad_revenue  2022-02-06 17:15:38       1644156938   

  event_receive_datetime  ...   ads_ltv   ecpm_sma normalized_ecpm_sma  \
0    2022-02-06 17:58:19  ...  0.487877   0.511508            0.511508   
1    2022-02-06 17:58:19  ...  0.488387   0.511389            0.511389   
2    2022-02-06 16:59:29  ...  0.678479   0.928597            0.928597   
3    2022-02-06 17:09:24  ...  0.013894  13.782633           13.893784   
4    2022-02-06 17:16:16  ...  0.060571   1.660282            1.660282   

      value currency ad_country_code       network_name        ad_unit_id  \
0  0.000510      USD              US       Google AdMob  64c6b1114ea8c1cc   
1  0.000510      USD              US       Google AdMob  64c6b1114ea8c1cc   
2  0.002386      USD              US  APPLOVIN_EXCHANGE  64c6b1114ea8c1cc   
3  0.013894      USD              US           AppLovin  3ab7705b07ebb75f   
4  0.006828      USD              US       Google AdMob  64c6b1114ea8c1cc   

                     ad_network_placement                  creativeId  
0  ca-app-pub-7511632782617496/8825263691  CJ3XirfD5vUCFYPnKAUdG3MOYQ  
1  ca-app-pub-7511632782617496/8825263691  CNa-lL_D5vUCFZrpKAUdU2QBDQ  
2                          banner_regular                  pandora-58  
3                           inter_regular                    13045364  
4  ca-app-pub-7511632782617496/6197324251  CMTi5e-h6_UCFUdVwQodt9EIMw  

[5 rows x 32 columns]

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 860079 entries, 0 to 860078
Data columns (total 32 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   profile_id               860065 non-null  object 
 1   os_version               860079 non-null  object 
 2   device_manufacturer      860079 non-null  object 
 3   device_model             860079 non-null  object 
 4   device_type              860079 non-null  object 
 5   device_locale            860079 non-null  object 
 6   event_name               860079 non-null  object 
 7   event_datetime           860079 non-null  object 
 8   event_timestamp          860079 non-null  int64  
 9   event_receive_datetime   860079 non-null  object 
 10  event_receive_timestamp  860079 non-null  int64  
 11  connection_type          860079 non-null  object 
 12  operator_name            841200 non-null  object 
 13  country_iso_code         860079 non-null  object 
 14  city

In [6]:
data['city'] = data['city'].replace(np.nan, 'missing')

# Functions

In [7]:
def data_grouping(data, group):
    df = data.groupby(["ad_type",group])['revenue'].apply(np.mean)
    df = df.to_frame().reset_index().sort_values(by="revenue", ascending=False)
    
    df["revenue"] = (df["revenue"]*1000)
    df = df.rename(columns={"revenue": "eCPM"})
    
    return df

In [8]:
def interactive_visualizations(data, col_name, graph_type, y_value):
    if graph_type == 'hist':
        def hist(chooser):
            for i in col_name:
                if chooser == i:
                    fig = px.histogram(data[data['ad_type'] == i],
                                       x=y_value, nbins=100)
                    fig.show()
            return
        print(interact(hist, chooser = col_name))

    elif graph_type == 'table':
        def table(chooser):
            for i in col_name:
                if chooser == i:
                    print(data[i].value_counts())
            return
        print(interact(table, chooser = col_name))
    
    elif graph_type == 'bar':
        def series(chooser):
            for i in col_name[:-1]:
                if chooser == i:
                    fig = px.bar(data[int(i.split()[1]) - 1 :int(i.split()[-1])],
                                 x='city', y=y_value)
                    fig.show()
                    
            return
        print(interact(series, chooser = col_name[:-1]))
    
    elif graph_type == 'os_bar':
        def mean_os(chooser):
            for i in col_name:
                if chooser == i:
                    fig = px.bar(data[data['ad_type'] == i],
                                 x='os_version',
                                 y=y_value)
                    fig.show()
            return
        print(interact(mean_os, chooser = col_name))
    
        
    

In [9]:
def conditions(event_datetime):
    event_datetime = int(event_datetime.split()[1].split(':')[0])
    
    if event_datetime >= 5 and event_datetime < 12:
        return "morning"
    elif event_datetime >= 12 and event_datetime < 17:
        return "afternoon"
    elif event_datetime >= 17 and event_datetime < 21:
        return "evening"
    else:
        return "night"

# 

# Tasks

In [11]:
col = data.columns
col_name = []
for i in col:
        col_name.append(i)
interactive_visualizations(data, col_name, 'table', 'revenue')

interactive(children=(Dropdown(description='chooser', options=('profile_id', 'os_version', 'device_manufacture…

<function interactive_visualizations.<locals>.table at 0x7f80c0068e50>


# 

## Task 1.a

For each ad type show the distribution of average user eCPM

In [12]:
mean_user = data_grouping(data, 'profile_id')
mean_user

ad_type   profile_id        eCPM
8254      rv  JLtY7hvRXO3  808.294820
7505      rv  JL5Tof1j4Ek  693.534973
8345      rv  JLw6kvDTnNC  601.758453
7709      rv  JLfKsX7jRRX  597.984497
5383     int  JLs8H01mzVe  568.678270
...      ...          ...         ...
394   banner  JLRcvaG98Uv    0.027756
782   banner  JLk8mMdNtMK    0.025439
2038  banner  JLv4xkSXmUs    0.022929
57    banner  JL3xP8xM4Xf    0.013930
3656     int  JL7PaV98HXg    0.000000

[8794 rows x 3 columns]

In [13]:
col_name = ["banner","int","rv"]
interactive_visualizations(mean_user, col_name, 'hist', 'eCPM')

interactive(children=(Dropdown(description='chooser', options=('banner', 'int', 'rv'), value='banner'), Output…

<function interactive_visualizations.<locals>.hist at 0x7f807f866a60>


From the presented graph we may observe the following data insights:
The average user eCPM metric is in range:
1. (0,10) for banners
2. (0,200) for int
3. (0,220) for rv

# 

## Task 1.b

For each ad type show the distribution of average eCPM per country & OS version

## Per city distribution

In [14]:
mean_city = data_grouping(data, "city")
col_name_city = []
ad_type = ["banner","int","rv"]
for j in ad_type:
    l = []
    for i in range(0, len(mean_city[mean_city['ad_type'] == j]) , 20):
        l.append(f'Top {i+1} - {i+20}')
    col_name_city.append(l)

### Banner

In [15]:
interactive_visualizations(mean_city[mean_city['ad_type'] == 'banner'], 
                           col_name_city[0], 'bar', 'eCPM')

interactive(children=(Dropdown(description='chooser', options=('Top 1 - 20', 'Top 21 - 40', 'Top 41 - 60', 'To…

<function interactive_visualizations.<locals>.series at 0x7f80bc4e6af0>


## INT 

In [16]:
interactive_visualizations(mean_city[mean_city['ad_type'] == 'int'], 
                           col_name_city[1], 'bar', 'eCPM')

interactive(children=(Dropdown(description='chooser', options=('Top 1 - 20', 'Top 21 - 40', 'Top 41 - 60', 'To…

<function interactive_visualizations.<locals>.series at 0x7f807f8669d0>


## RV 

In [17]:
interactive_visualizations(mean_city[mean_city['ad_type'] == 'rv'], 
                           col_name_city[2], 'bar', 'eCPM')

interactive(children=(Dropdown(description='chooser', options=('Top 1 - 20', 'Top 21 - 40', 'Top 41 - 60', 'To…

<function interactive_visualizations.<locals>.series at 0x7f80bd5cfe50>


                    Top 3 cities for each advertising type
Banner
1. La Crosse
2. Mountain View
3. Huron

INT
1. Madisonville
2. La Crosse
3. Rohnert Park

RV
1. East Peoria
2. Norwood
3. Idaho Falls

### 

## Per OS version distribution

In [18]:
col_name = ['banner', 'int', 'rv']

mean_os = data_grouping(data, "os_version")
interactive_visualizations(mean_os, col_name, 'os_bar', 'eCPM')

interactive(children=(Dropdown(description='chooser', options=('banner', 'int', 'rv'), value='banner'), Output…

<function interactive_visualizations.<locals>.mean_os at 0x7f80bc4e6ee0>


                    Top 3 os versions for each advertising type
Banner
1. "12"
2. "11"
3. "10"

INT
1. "12"
2. "7"
3. "11"

RV
1. "7"
2. "12"
3. "11"

## Task 1.c

For each ad type show the sum of revenue per country & OS version

## Per OS version

In [19]:
col_name = ['banner', 'int', 'rv']
sum_os = data.groupby(["ad_type",'os_version'])['revenue'].apply(np.sum)
sum_os = sum_os.to_frame().reset_index().sort_values(by="revenue", ascending=False)
interactive_visualizations(sum_os, col_name, 'os_bar', 'revenue')

interactive(children=(Dropdown(description='chooser', options=('banner', 'int', 'rv'), value='banner'), Output…

<function interactive_visualizations.<locals>.mean_os at 0x7f80bc4e6c10>


Top 3 profitable os versions for each advertising type are

1. "12"
2. "11"
3. "10"

## Per city 

missing -> revenue from advertisings with missing city information

### Banner 

In [35]:
sum_city = data.groupby(["ad_type",'city'])['revenue'].apply(np.sum)
sum_city = sum_city.to_frame().reset_index().sort_values(by="revenue", ascending=False)
interactive_visualizations(sum_city[sum_city['ad_type'] == 'banner'], 
                           col_name_city[0], 'bar', 'revenue')

interactive(children=(Dropdown(description='chooser', options=('Top 1 - 20', 'Top 21 - 40', 'Top 41 - 60', 'To…

<function interactive_visualizations.<locals>.series at 0x7f8061b0cb80>


### INT

In [22]:
interactive_visualizations(sum_city[sum_city['ad_type'] == 'int'], 
                           col_name_city[1], 'bar', 'revenue')

interactive(children=(Dropdown(description='chooser', options=('Top 1 - 20', 'Top 21 - 40', 'Top 41 - 60', 'To…

<function interactive_visualizations.<locals>.series at 0x7f8061af2820>


### RV

In [23]:
interactive_visualizations(sum_city[sum_city['ad_type'] == 'rv'], 
                           col_name_city[2], 'bar', 'revenue')

interactive(children=(Dropdown(description='chooser', options=('Top 1 - 20', 'Top 21 - 40', 'Top 41 - 60', 'To…

<function interactive_visualizations.<locals>.series at 0x7f8061af2b80>


                    Top 3 cities for each advertising type
Banner
1. "missing"
2. New York
3. Indianapolis

INT
1. "missing"
2. New York
3. Indianapolis

RV
1. "missing"
2. Houston
3. Minneapolis

### 

## Task 2 

With 95% confidence level and 5% error confidence interval, for given population
the minimum required sample size equals to 363. Based on the latter, we may remove
cities which have less than 363 rows of data.

Afterwards, we will check again the average eCPM metric for each ad type per countries.

In [24]:
filtered_data = data.groupby(['ad_type', 'city']).agg({'profile_id': ['count'], 'revenue':['mean']})
filtered_data[('revenue', 'mean')] *= 1000

filtered_data = filtered_data[filtered_data[('profile_id', 'count')] >= 363
                             ].reset_index().sort_values(
                                       by=('revenue', 'mean'), ascending=False)

filtered_data.columns = filtered_data.columns.droplevel(1)
filtered_data = filtered_data.rename(columns={'profile_id': "count",
                                              'revenue': "eCPM"})

In [25]:
col_name_city = []
ad_type = ["banner","int","rv"]
for j in ad_type:
    l = []
    for i in range(0, len(filtered_data[filtered_data['ad_type'] == j]) , 20):
        l.append(f'Top {i+1} - {i+20}')
    col_name_city.append(l)

### Banner 

In [26]:
interactive_visualizations(filtered_data[filtered_data['ad_type'] == 'banner'], 
                           col_name_city[0], 'bar', 'eCPM')

interactive(children=(Dropdown(description='chooser', options=('Top 1 - 20', 'Top 21 - 40', 'Top 41 - 60', 'To…

<function interactive_visualizations.<locals>.series at 0x7f80616cb820>


### INT 

In [27]:
fig = px.bar(filtered_data[filtered_data['ad_type'] == 'int'],
                                 x='city', y='eCPM')
fig.show()

### RV 

In [28]:
fig = px.bar(filtered_data[filtered_data['ad_type'] == 'rv'],
                                 x='city', y='eCPM')
fig.show()

                                  Data insights

After the removal of cities which do not have enough data to come up with statistically significant conclusions, we may see that the overall results of leading cities by average eCPM metric have changed for all the advertising types. 

Besides, we could also remove the data concerning "missing" cities, as they do not contain any significant information.

# Task 4.1

What percentage of revenue contains information about users' city?

In [46]:
city_revenue = sum_city.groupby(["city"])['revenue'].sum()
sum_missing = city_revenue.iloc[-1]
sum_missing

870.8460633825665

In [48]:
sum_not_missing = city_revenue[:-1].sum()
sum_not_missing

1519.9525969609954

In [49]:
labels = ['Missing','Not missing']
values = [sum_missing, sum_not_missing]

fig = go.Figure(data=[go.Pie(labels=labels, values=values)])
fig.show()

About 36,4% of revenue is coming from users with not mentioned city names.

# Task 4.2

Does the eCPM metric depend on the time of the day?

In [50]:
time_data = data[['event_datetime', 'revenue']]
time_data.head()

event_datetime   revenue
0  2022-02-04 20:02:28  0.000510
1  2022-02-04 20:02:45  0.000510
2  2022-02-05 17:53:05  0.002386
3  2022-02-06 17:09:01  0.013894
4  2022-02-06 17:15:38  0.006828

In [51]:
time_list = []
for i in time_data['event_datetime']:
    time_list.append(conditions(i))
    
time_data.insert(loc=2, column='time_category', value=time_list)

time_data.head()

event_datetime   revenue time_category
0  2022-02-04 20:02:28  0.000510       evening
1  2022-02-04 20:02:45  0.000510       evening
2  2022-02-05 17:53:05  0.002386       evening
3  2022-02-06 17:09:01  0.013894       evening
4  2022-02-06 17:15:38  0.006828       evening

In [52]:
mean_time_category = time_data.groupby(["time_category"])['revenue'].mean()
mean_time_category = pd.DataFrame(mean_time_category * 1000).sort_values(by="revenue",
                                                              ascending=False)
mean_time_category

revenue
time_category          
evening        2.898593
morning        2.877293
afternoon      2.790969
night          2.662209

In [53]:
labels = mean_time_category.index
values = mean_time_category['revenue']

fig = go.Figure(data=[go.Pie(labels=labels, values=values)])
fig.show()

                               Data insights
                               
Overall, revenue does not have strong dependency from the time of day, though the highest revenue is generated in the evening & morning.

## Conclusion

1. The leading OS versions by average eCPM are "12", "11", "10" & "7"

2. Top 3 OS versions generating the highest revenue are "12", "11" & "10"

3. Leading cities generating highest revenue are New York, Indianapolis, Houston & Minneapolis

4. After the removal of cities with not sufficient data, the list of cities with highest average eCPM metric is:
   Banner
   1) Clarks Summit
   2) Gardena
   3) Waynesboro
   
   INT
   1) New York
   2) Houston
   3) Charlotte
   
   RV
   1) contains data with missing city names only
   
   
5. Overall, 36,4% of generated revenue is coming from unknown cities (870), while the rest of revenue (1520) is coming from certain cities.